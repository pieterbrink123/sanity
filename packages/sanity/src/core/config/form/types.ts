import {type ComponentType} from 'react'

import {type PreviewProps} from '../../components/previews'
import {type FieldProps, type InputProps, type ItemProps} from '../../form/types'
import {type BlockAnnotationProps, type BlockProps} from '../../form/types/blockProps'

/**
 * @hidden
 * @beta */
export interface FormComponents {
  annotation?: ComponentType<BlockAnnotationProps>
  block?: ComponentType<BlockProps>
  field?: ComponentType<FieldProps>
  inlineBlock?: ComponentType<BlockProps>
  input?: ComponentType<InputProps>
  item?: ComponentType<ItemProps>
  preview?: ComponentType<PreviewProps>
}
