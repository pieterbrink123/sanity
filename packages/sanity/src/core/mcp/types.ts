import {type EditorSelection} from '@portabletext/editor'
import {type Path, type SanityDocumentLike} from '@sanity/types'
import {type ComponentType, type ReactNode} from 'react'
/**
 * @internal
 * @hidden
 */
export interface MCPAgentContext {
  list?: TODO
  presence?: TODO
  focus?: {
    path?: Path
    selection?: EditorSelection
    document?: SanityDocumentLike
  }
}

/**
 * @internal
 * @hidden
 */
export type MCPComponentProps = {
  children?: ReactNode
  active?: boolean
  onDeactivate?: () => void
  context: MCPAgentContext
}

/**
 * @internal
 * @hidden
 */
export interface MCPConfig {
  component: ComponentType<MCPComponentProps>
}

/**
 * @internal
 * @hidden
 */
export type TODO = never

export type MCPFocusEvent = {
  type: 'UPDATE_FOCUS'
  path?: Path
  selection?: EditorSelection
  document?: SanityDocumentLike
}

/**
 * @internal
 * @hidden
 */
export type MCPEvent = MCPFocusEvent

/**
 * @internal
 * @hidden
 */
export interface MCPProviderContextValue {
  onEvent: (mcpEvent: MCPEvent) => void
}
