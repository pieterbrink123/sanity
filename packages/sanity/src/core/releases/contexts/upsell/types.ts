import {type UpsellDialogViewedInfo} from '../../../studio'
import {type UpsellData} from '../../../studio/upsell/types'

export interface ReleasesUpsellContextValue {
  upsellDialogOpen: boolean
  handleOpenDialog: (source: UpsellDialogViewedInfo['source']) => void
  execIfNotUpsell: (callback: () => void) => Promise<boolean>
  upsellData: UpsellData | null
  telemetryLogs: {
    dialogSecondaryClicked: () => void
    dialogPrimaryClicked: () => void
    panelViewed: (source: UpsellDialogViewedInfo['source']) => void
    panelDismissed: () => void
    panelPrimaryClicked: () => void
    panelSecondaryClicked: () => void
  }
}
