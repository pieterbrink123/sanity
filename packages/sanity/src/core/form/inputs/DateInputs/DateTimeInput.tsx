/* eslint-disable no-restricted-imports */
import {Box, Card, Flex, Inline} from '@sanity/ui'
import {
  DEFAULT_DATE_FORMAT,
  DEFAULT_TIME_FORMAT,
  format,
  isValidTimeZoneString,
  parse,
  type ParseResult,
} from '@sanity/util/legacyDateFormat'
import {getMinutes, parseISO, setMinutes} from 'date-fns'
import {useCallback, useMemo} from 'react'
import {
  ChangeIndicator,
  EMPTY_ARRAY,
  FormFieldHeaderText,
  FormFieldStatus,
  set,
  type StringInputProps,
  unset,
  useFieldActions,
} from 'sanity'
import styled from 'styled-components'

import {type CalendarLabels} from '../../../components/inputs/DateInputs/calendar/types'
import {useTranslation} from '../../../i18n'
import ButtonTimeZone from '../../../scheduledPublishing/components/timeZoneButton/TimeZoneButton'
import ButtonTimeZoneElementQuery from '../../../scheduledPublishing/components/timeZoneButton/TimeZoneButtonElementQuery'
import useTimeZone, {TimeZoneScopeType} from '../../../scheduledPublishing/hooks/useTimeZone'
import {FormFieldBaseHeader} from '../../components/formField/FormFieldBaseHeader'
import {CommonDateTimeInput} from './CommonDateTimeInput'
import {getCalendarLabels, isValidDate} from './utils'

const Root = styled(Card)`
  line-height: 1;
`

const CenterAlignedBox = styled(Box)`
  align-self: center;
`

const ZeroLineHeightBox = styled(Box)`
  line-height: 0;
`

interface ParsedOptions {
  dateFormat: string
  timeFormat: string
  timeStep: number
  displayTimeZone?: string
  allowTimeZoneSwitch?: boolean
}

interface SchemaOptions {
  dateFormat?: string
  timeFormat?: string
  timeStep?: number
  displayTimeZone?: string
  allowTimeZoneSwitch?: boolean
}

/**
 * @hidden
 * @beta */
export type DateTimeInputProps = StringInputProps

const serialize = (date: Date) => {
  return date.toISOString()
}

function parseOptions(options: SchemaOptions = {}): ParsedOptions {
  return {
    dateFormat: options.dateFormat || DEFAULT_DATE_FORMAT,
    timeFormat: options.timeFormat || DEFAULT_TIME_FORMAT,
    timeStep: ('timeStep' in options && Number(options.timeStep)) || 1,
    displayTimeZone:
      options.displayTimeZone && isValidTimeZoneString(options.displayTimeZone)
        ? options.displayTimeZone
        : undefined,
    allowTimeZoneSwitch:
      options.allowTimeZoneSwitch === undefined ? true : options.allowTimeZoneSwitch,
  }
}

const getDeserializer =
  (timeZoneName?: string) =>
  (isoString: string): ParseResult => {
    // create a date object respecting the time zone
    const {date: deserialized} = parse(isoString, undefined, timeZoneName)
    if (deserialized && isValidDate(deserialized)) {
      return {isValid: true, date: deserialized}
    }
    return {isValid: false, error: `Invalid date value: "${isoString}"`}
  }

// enforceTimeStep takes a dateString and datetime schema options and enforces the time
// to be within the configured timeStep
function enforceTimeStep(dateString: string, timeStep: number) {
  if (!timeStep || timeStep === 1) {
    return dateString
  }

  const date = parseISO(dateString)
  const minutes = getMinutes(date)
  const leftOver = minutes % timeStep
  if (leftOver !== 0) {
    return serialize(setMinutes(date, minutes - leftOver))
  }

  return serialize(date)
}

/**
 * @hidden
 * @beta */
export function DateTimeInput(props: DateTimeInputProps) {
  const {
    onChange,
    schemaType,
    value,
    elementProps,
    id,
    validation,
    changed,
    path,
    presence = EMPTY_ARRAY,
  } = props

  const {
    focused,
    __internal_comments: comments,
    hovered,
    onMouseEnter,
    onMouseLeave,
    actions,
    __internal_slot: slot,
  } = useFieldActions()
  const {
    dateFormat,
    timeFormat,
    timeStep,
    displayTimeZone,
    allowTimeZoneSwitch = true,
  } = parseOptions(schemaType.options)
  const {t} = useTranslation()
  const timeZoneScope = useMemo(
    () => ({type: TimeZoneScopeType.input, defaultTimeZone: displayTimeZone, id}),
    [displayTimeZone, id],
  )

  const {timeZone} = useTimeZone(timeZoneScope)

  const handleChange = useCallback(
    (nextDate: string | null) => {
      let date = nextDate
      if (date !== null && timeStep > 1) {
        date = enforceTimeStep(date, timeStep)
      }

      onChange(date === null ? unset() : set(date))
    },
    [onChange, timeStep],
  )

  const formatInputValue = useCallback(
    (date: Date) => format(date, `${dateFormat} ${timeFormat}`, {timeZone: timeZone.name}),
    [dateFormat, timeFormat, timeZone.name],
  )

  const deserialize = useMemo(() => getDeserializer(timeZone.name), [timeZone.name])
  const memoizedSerialize = useMemo(() => serialize, [])

  const parseInputValue = useCallback(
    (inputValue: string): ParseResult =>
      parse(inputValue, `${dateFormat} ${timeFormat}`, timeZone.name),
    [dateFormat, timeFormat, timeZone.name],
  )
  const calendarLabels: CalendarLabels = useMemo(() => getCalendarLabels(t), [t])
  const commonProps = useMemo(
    () => ({
      ...elementProps,
      calendarLabels,
      deserialize,
      formatInputValue,
      onChange: handleChange,
      parseInputValue,
      placeholder: schemaType.placeholder,
      serialize: memoizedSerialize,
      timeZone: timeZone.name,
      timeStep,
      selectTime: true,
      value,
      timeZoneScope,
    }),
    [
      elementProps,
      calendarLabels,
      deserialize,
      formatInputValue,
      handleChange,
      parseInputValue,
      schemaType.placeholder,
      memoizedSerialize,
      timeZone.name,
      timeStep,
      value,
      timeZoneScope,
    ],
  )

  const input = (
    <>
      <CommonDateTimeInput {...commonProps} />
    </>
  )
  return (
    <Root
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      data-testid="datetime-input"
      radius={2}
    >
      <Flex direction={'column'}>
        <Box flex={1} paddingY={2}>
          <FormFieldBaseHeader
            __internal_comments={comments}
            __internal_slot={slot}
            actions={actions}
            fieldFocused={Boolean(focused)}
            fieldHovered={hovered}
            presence={presence}
            inputId={id}
            content={
              <Inline>
                <FormFieldHeaderText
                  deprecated={schemaType.deprecated}
                  description={schemaType.description}
                  inputId={id}
                  validation={validation}
                  title={schemaType.title}
                />
                {displayTimeZone && (
                  <ButtonTimeZoneElementQuery>
                    <Box marginLeft={2}>
                      <ButtonTimeZone
                        allowTimeZoneSwitch={allowTimeZoneSwitch}
                        useElementQueries
                        timeZoneScope={timeZoneScope}
                      />
                    </Box>
                  </ButtonTimeZoneElementQuery>
                )}
              </Inline>
            }
          />
        </Box>
        <CenterAlignedBox paddingX={3} paddingY={1}>
          <FormFieldStatus maxAvatars={1} position="top">
            {/*<FieldPresence maxAvatars={1} presence={presence} />*/}
          </FormFieldStatus>
        </CenterAlignedBox>
        <ChangeIndicator hasFocus={Boolean(focused)} isChanged={changed} path={path}>
          {input}
        </ChangeIndicator>
      </Flex>
    </Root>
  )
}
