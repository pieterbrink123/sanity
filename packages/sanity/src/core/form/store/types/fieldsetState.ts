import {type Path} from '@sanity/types'
import {type Columns, type ResponsiveProp} from '@sanity/ui/css'

import {type FieldError} from './memberErrors'
import {type FieldMember} from './members'

/**
 * @hidden
 * @beta */
export interface FieldsetState {
  path: Path
  name: string
  level: number
  title?: string
  description?: string
  hidden?: boolean
  collapsible?: boolean
  collapsed?: boolean
  columns?: ResponsiveProp<Columns>
  members: (FieldMember | FieldError)[]
}
