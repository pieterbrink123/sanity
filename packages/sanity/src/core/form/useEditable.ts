import {type ReleaseId, type SanityDocument} from '@sanity/client'
import {isActionEnabled} from '@sanity/schema/_internal'
import {useTelemetry} from '@sanity/telemetry/react'
import {
  type ObjectSchemaType,
  type Path,
  type SanityDocumentLike,
  type ValidationMarker,
} from '@sanity/types'
import {useToast} from '@sanity/ui'
import {pathFor} from '@sanity/util/paths'
import {throttle} from 'lodash'
import {
  type RefObject,
  useCallback,
  useEffect,
  useInsertionEffect,
  useMemo,
  useRef,
  useState,
} from 'react'

import {isSanityCreateLinkedDocument} from '../create/createUtils'
import {type ConnectionState, useConnectionState} from '../hooks/useConnectionState'
import {useDocumentOperation} from '../hooks/useDocumentOperation'
import {useEditState} from '../hooks/useEditState'
import {useSchema} from '../hooks/useSchema'
import {useValidationStatus} from '../hooks/useValidationStatus'
import {useTranslation} from '../i18n/hooks/useTranslation'
import {
  type DocumentPresence,
  type EditStateFor,
  type PermissionCheckResult,
  useDocumentValuePermissions,
  usePresenceStore,
} from '../store'
import {EMPTY_ARRAY, useUnique} from '../util'
import {deepEquals} from '../validation/util/deepEquals'
import {
  type FormState,
  getExpandOperations,
  type OnPathFocusPayload,
  type PatchEvent,
  setAtPath,
  type StateTree,
  toMutationPatches,
  useFormState,
} from '.'
import {CreatedDraft} from './__telemetry__/form.telemetry'

interface EditableOptions {
  documentType: string
  documentId: string
  initialValue?: SanityDocumentLike
  releaseId?: ReleaseId
  initialFocusPath?: Path
  selectedPerspectiveName?: ReleaseId | 'published'
  readOnly?: boolean | ((editState: EditStateFor) => boolean)
  /**
   * Usually the historical _rev value selected, if not defined, it will use the current document value
   * so no comparison will be done.
   */
  comparisonValue?:
    | Partial<SanityDocument>
    | ((editState: EditStateFor) => Partial<SanityDocument>)
    | null
  onFocusPath?: (path: Path) => void
}
interface EditableValues {
  editState: EditStateFor
  connectionState: ConnectionState
  collapsedFieldSets: StateTree<boolean> | undefined
  collapsedPaths: StateTree<boolean> | undefined
  openPath: Path
  onBlur: (blurredPath: Path) => void
  onFocus: (_nextFocusPath: Path, payload?: OnPathFocusPayload) => void
  onSetCollapsedPath: (path: Path, collapsed: boolean) => void
  onSetActiveFieldGroup: (path: Path, groupName: string) => void
  onSetCollapsedFieldSet: (path: Path, collapsed: boolean) => void
  ready: boolean
  value: SanityDocumentLike
  formState: FormState
  focusPath: Path
  validation: ValidationMarker[]
  permissions: PermissionCheckResult | undefined
  isPermissionsLoading: boolean
  onChange: (event: PatchEvent) => void
  onPathOpen: (path: Path) => void
  onProgrammaticFocus: (nextPath: Path) => void
  formStateRef: RefObject<FormState>
  schemaType: ObjectSchemaType
}

export function useEditable(options: EditableOptions): EditableValues {
  const {
    documentType,
    documentId,
    initialValue,
    comparisonValue: comparisonValueRaw,
    releaseId,
    initialFocusPath,
    selectedPerspectiveName,
    readOnly: readOnlyProp,
    onFocusPath,
  } = options
  const schema = useSchema()
  const presenceStore = usePresenceStore()

  const schemaType = schema.get(documentType) as ObjectSchemaType | undefined
  if (!schemaType) {
    throw new Error(`Schema type for '${documentType}' not found`)
  }
  const telemetry = useTelemetry()

  const {validation: validationRaw} = useValidationStatus(documentId, documentType, releaseId)
  const validation = useUnique(validationRaw)

  const editState = useEditState(documentId, documentType, 'default', releaseId)
  const connectionState = useConnectionState(documentId, documentType, {version: releaseId})
  useConnectionToast(connectionState)

  const [focusPath, setFocusPath] = useState<Path>(initialFocusPath || EMPTY_ARRAY)
  const [openPath, setOpenPathState] = useState<Path>(EMPTY_ARRAY)
  const [collapsedPaths, onSetCollapsedPath] = useState<StateTree<boolean>>()
  const [collapsedFieldSets, onSetCollapsedFieldSets] = useState<StateTree<boolean>>()
  const [fieldGroupState, onSetFieldGroupState] = useState<StateTree<string>>()
  const [presence, setPresence] = useState<DocumentPresence[]>([])

  const handleOnSetCollapsedPath = useCallback((path: Path, collapsed: boolean) => {
    onSetCollapsedPath((prevState) => setAtPath(prevState, path, collapsed))
  }, [])

  const handleOnSetCollapsedFieldSet = useCallback((path: Path, collapsed: boolean) => {
    onSetCollapsedFieldSets((prevState) => setAtPath(prevState, path, collapsed))
  }, [])

  const handleSetActiveFieldGroup = useCallback(
    (path: Path, groupName: string) =>
      onSetFieldGroupState((prevState) => setAtPath(prevState, path, groupName)),
    [],
  )

  const comparisonValue = useMemo(() => {
    if (typeof comparisonValueRaw === 'function') {
      return comparisonValueRaw(editState)
    }
    return comparisonValueRaw
  }, [comparisonValueRaw, editState])

  const {patch} = useDocumentOperation(documentId, documentType)

  const patchRef = useRef<(event: PatchEvent) => void>(() => {
    throw new Error(
      'Attempted to patch the Sanity document during initial render or in an `useInsertionEffect`. Input components should only call `onChange()` in a useEffect or an event handler.',
    )
  })
  useInsertionEffect(() => {
    // note: this needs to happen in an insertion effect to make sure we're ready to receive patches from child components when they run their effects initially
    // in case they do e.g. `useEffect(() => props.onChange(set("foo")), [])`
    // Note: although we discourage patch-on-mount, we still support it.
    patchRef.current = (event: PatchEvent) => {
      // when creating a new draft
      if (!editState.draft && !editState.published && !editState.version) {
        telemetry.log(CreatedDraft)
      }
      patch.execute(toMutationPatches(event.patches), initialValue)
    }
  }, [editState.draft, editState.published, editState.version, initialValue, patch, telemetry])

  const handleChange = useCallback((event: PatchEvent) => patchRef.current(event), [])

  const documentValue: SanityDocumentLike = useMemo(() => {
    const baseValue = initialValue || {
      _id: documentId,
      _type: documentType,
    }
    if (releaseId) {
      return (
        editState.version || editState.draft || editState.published || initialValue || baseValue
      )
    }
    if (selectedPerspectiveName === 'published') {
      return editState.published || editState.draft || initialValue || baseValue
    }
    return editState?.draft || editState?.published || initialValue || baseValue
  }, [
    documentId,
    documentType,
    editState.draft,
    editState.published,
    editState.version,
    initialValue,
    releaseId,
    selectedPerspectiveName,
  ])

  useEffect(() => {
    const subscription = presenceStore
      .documentPresence(documentValue._id)
      .subscribe((nextPresence) => {
        setPresence(nextPresence)
      })
    return () => {
      subscription.unsubscribe()
    }
  }, [presenceStore, documentValue._id])

  const [permissions, isPermissionsLoading] = useDocumentValuePermissions({
    document: documentValue,
    permission: documentValue._createdAt ? 'update' : 'create',
  })

  const isNonExistent = !documentValue?._id
  const isCreateLinked = isSanityCreateLinkedDocument(documentValue)
  const ready = editState.ready && connectionState === 'connected'

  const readOnly = useMemo(() => {
    const createActionDisabled = isNonExistent && !isActionEnabled(schemaType!, 'create')
    const updateActionDisabled = !isActionEnabled(schemaType!, 'update')

    const hasNoPermission = !isPermissionsLoading && !permissions?.granted
    const reconnecting = connectionState === 'reconnecting'
    const isLocked = editState.transactionSyncLock?.enabled

    const editableReadOnly =
      reconnecting ||
      hasNoPermission ||
      isLocked ||
      createActionDisabled ||
      updateActionDisabled ||
      isCreateLinked ||
      !ready

    if (editableReadOnly) return true
    if (typeof readOnlyProp === 'function') return readOnlyProp(editState)
    return Boolean(readOnlyProp)
  }, [
    isNonExistent,
    schemaType,
    isPermissionsLoading,
    permissions?.granted,
    connectionState,
    ready,
    editState,
    isCreateLinked,
    readOnlyProp,
  ])

  const formState = useFormState({
    schemaType,
    documentValue,
    comparisonValue: comparisonValue || documentValue,
    readOnly,
    changesOpen: false,
    presence,
    focusPath,
    openPath,
    collapsedPaths,
    collapsedFieldSets,
    fieldGroupState,
    validation,
  })!

  const formStateRef = useRef(formState)
  useEffect(() => {
    formStateRef.current = formState
  }, [formState])

  const setOpenPath = useCallback(
    (path: Path) => {
      const ops = getExpandOperations(formStateRef.current!, path)
      ops.forEach((op) => {
        if (op.type === 'expandPath') {
          onSetCollapsedPath((prevState) => setAtPath(prevState, op.path, false))
        }
        if (op.type === 'expandFieldSet') {
          onSetCollapsedFieldSets((prevState) => setAtPath(prevState, op.path, false))
        }
        if (op.type === 'setSelectedGroup') {
          onSetFieldGroupState((prevState) => setAtPath(prevState, op.path, op.groupName))
        }
      })
      setOpenPathState(path)
    },
    [formStateRef],
  )

  const updatePresence = useCallback(
    (nextFocusPath: Path, payload?: OnPathFocusPayload) => {
      presenceStore.setLocation([
        {
          type: 'document',
          documentId: documentValue._id,
          path: nextFocusPath,
          lastActiveAt: new Date().toISOString(),
          selection: payload?.selection,
        },
      ])
    },
    [presenceStore, documentValue._id],
  )

  const updatePresenceThrottled = useMemo(
    () => throttle(updatePresence, 1000, {leading: true, trailing: true}),
    [updatePresence],
  )
  const focusPathRef = useRef<Path>([])

  const handleFocus = useCallback(
    (_nextFocusPath: Path, payload?: OnPathFocusPayload) => {
      const nextFocusPath = pathFor(_nextFocusPath)
      if (nextFocusPath !== focusPathRef.current) {
        setFocusPath(pathFor(nextFocusPath))
        setOpenPath(pathFor(nextFocusPath.slice(0, -1)))
        focusPathRef.current = nextFocusPath
        onFocusPath?.(nextFocusPath)
      }
      updatePresenceThrottled(nextFocusPath, payload)
    },
    [onFocusPath, setFocusPath, setOpenPath, updatePresenceThrottled],
  )

  const disableBlurRef = useRef(false)

  const handleBlur = useCallback(
    (_blurredPath: Path) => {
      if (disableBlurRef.current) {
        return
      }

      setFocusPath(EMPTY_ARRAY)

      if (focusPathRef.current !== EMPTY_ARRAY) {
        focusPathRef.current = EMPTY_ARRAY
        onFocusPath?.(EMPTY_ARRAY)
      }

      // note: we're deliberately not syncing presence here since it would make the user avatar disappear when a
      // user clicks outside a field without focusing another one
    },
    [onFocusPath, setFocusPath],
  )

  const handleProgrammaticFocus = useCallback(
    (nextPath: Path) => {
      // Supports changing the focus path not by a user interaction, but by a programmatic change, e.g. the url path changes.
      // to avoid the blur event to be triggered, we set a flag to disable it for a short period of time.
      disableBlurRef.current = true

      // Reset focus path when url params path changes
      if (!deepEquals(focusPathRef.current, nextPath)) {
        setFocusPath(nextPath)
        setOpenPath(nextPath)
        onFocusPath?.(nextPath)

        focusPathRef.current = nextPath
      }

      const timeout = setTimeout(() => {
        disableBlurRef.current = false
      }, 0)
      return () => clearTimeout(timeout)
    },
    [onFocusPath, setOpenPath],
  )
  return {
    editState,
    connectionState,
    focusPath,
    validation,
    ready,
    value: documentValue,
    formState,
    permissions,
    isPermissionsLoading,
    formStateRef,

    collapsedFieldSets,
    collapsedPaths,
    openPath,
    schemaType,
    onChange: handleChange,
    onPathOpen: setOpenPath,
    onProgrammaticFocus: handleProgrammaticFocus,
    onBlur: handleBlur,
    onFocus: handleFocus,
    onSetActiveFieldGroup: handleSetActiveFieldGroup,
    onSetCollapsedPath: handleOnSetCollapsedPath,
    onSetCollapsedFieldSet: handleOnSetCollapsedFieldSet,
  }
}

const useConnectionToast = (connectionState: ConnectionState) => {
  const {push: pushToast} = useToast()
  const {t} = useTranslation('studio')

  useEffect(() => {
    let timeout: ReturnType<typeof setTimeout>
    if (connectionState === 'reconnecting') {
      timeout = setTimeout(() => {
        pushToast({
          id: 'sanity/reconnecting',
          status: 'warning',
          title: t('panes.document-pane-provider.reconnecting.title'),
        })
      }, 2000) // 2 seconds, we can iterate on the value
    }
    return () => {
      if (timeout) clearTimeout(timeout)
    }
  }, [connectionState, pushToast, t])
}
