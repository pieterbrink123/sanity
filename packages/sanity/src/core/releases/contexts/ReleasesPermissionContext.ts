import {createContext} from 'sanity/_createContext'

export interface ReleasePermissionsValue {
  checkWithPermissionGuard: <T extends (...args: any[]) => Promise<void> | void>(
    action: T,
    ...args: Parameters<T>
  ) => Promise<boolean>
  permissions: Record<string, boolean>
}

/**
 * @internal
 */
export const ReleasesPermissionsContext = createContext<ReleasePermissionsValue | null>(
  'sanity/_singletons/context/releases-permissions',
  null,
)
