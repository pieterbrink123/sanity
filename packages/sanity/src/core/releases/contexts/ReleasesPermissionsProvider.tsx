import {useState} from 'react'

import {isErrorWithDetails} from '../../error/types/isErrorWithDetails'
import {ReleasesPermissionsContext} from './ReleasesPermissionContext'

type ReleasePermissionError = {details: {type: 'insufficientPermissionsError'}}

export const isReleasePermissionError = (error: unknown): error is ReleasePermissionError =>
  isErrorWithDetails(error) && error.details?.type === 'insufficientPermissionsError'

export const ReleasesPermissionsProvider = ({children}: {children: React.ReactNode}) => {
  const [permissions, setPermissions] = useState<Record<string, boolean>>({})

  /**
   * Checks if a release action can be performed by running a dry run of the given action
   *
   * @param action - any of the actions from the ReleaseOperationStore, e.g. publishRelease should send in also the needed props
   * @param args - the arguments to pass to the action (release id, etc)
   * @returns true or false depending if the user can perform the action
   */
  const checkWithPermissionGuard = async <T extends (...args: any[]) => Promise<void> | void>(
    action: T,
    ...args: Parameters<T>
  ): Promise<boolean> => {
    if (action.name in permissions) {
      return permissions[action.name]
    }

    try {
      await action(...args, {
        dryRun: true,
        skipCrossDatasetReferenceValidation: true,
      })
      setPermissions({...permissions, [action.name]: true})
      return true
    } catch (e) {
      setPermissions({...permissions, [action.name]: false})

      return !isReleasePermissionError(e)
    }
  }

  const context = {
    checkWithPermissionGuard: checkWithPermissionGuard,
    permissions: permissions,
  }
  // Avoid mounting the provider if it's already provided by a parent
  return (
    <ReleasesPermissionsContext.Provider value={context}>
      {children}
    </ReleasesPermissionsContext.Provider>
  )
}
