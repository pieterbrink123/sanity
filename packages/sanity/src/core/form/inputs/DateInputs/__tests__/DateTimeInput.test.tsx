import {defineField, type StringSchemaType} from '@sanity/types'
import {fireEvent} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {expect, test, vi} from 'vitest'

import {renderStringInput} from '../../../../../../test/form'
import {FormValueProvider} from '../../../contexts/FormValue'
import {type StringInputProps} from '../../../types'
import {DateTimeInput} from '../DateTimeInput'

vi.mock('sanity', () => ({
  set: vi.fn(),
}))

const DateTimeInputWithFormValue = (inputProps: StringInputProps<StringSchemaType>) => (
  <FormValueProvider value={{_id: 'test123', _type: 'datetime'}}>
    <DateTimeInput {...inputProps} />
  </FormValueProvider>
)

// NOTE: for the tests to be deterministic we need this to ensure tests are run in a predefined time zone
// see globalSetup in jest config for details about how this is set up
test('time zone for the test environment should be set to America/Los_Angeles', () => {
  expect(Intl.DateTimeFormat().resolvedOptions().timeZone).toBe('America/Los_Angeles')
})

test('does not emit onChange after invalid value has been typed', async () => {
  const {onChange, result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
    }),
    render: (inputProps) => <DateTimeInputWithFormValue {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  userEvent.type(input, 'this is invalid')
  expect(input.value).toBe('this is invalid')
  expect(onChange.mock.calls.length).toBe(0)

  fireEvent.blur(input)

  expect(onChange.mock.calls.length).toBe(0)
})

test('emits onChange on correct format if a valid value has been typed', async () => {
  const {onChange, result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
    }),
    render: (inputProps) => <DateTimeInputWithFormValue {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  // NOTE: the date is entered and displayed in local time zone
  // (which is hardcoded to America/Los_Angeles)
  userEvent.type(input, '2021-03-28 10:23')
  expect(input.value).toBe('2021-03-28 10:23')

  fireEvent.blur(input)

  // NOTE: the date is entered and displayed in local time zone but stored in utc
  expect(onChange.mock.calls).toMatchSnapshot()
})

test('formatting of deserialized value', async () => {
  const {result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
    }),
    props: {documentValue: {test: '2021-03-28T17:23:00.000Z'}},
    render: (inputProps) => <DateTimeInputWithFormValue {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  // const {textInput} = renderInput({value: '2021-03-28T17:23:00.000Z'} as any)
  expect(input.value).toBe('2021-03-28 10:23')
})

test('time is shown in the display time zone if specified (utc+1 winter)', async () => {
  const {result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
      options: {displayTimeZone: 'Europe/Oslo'},
    }),
    props: {documentValue: {test: '2021-01-15T12:00:00.000Z'}},
    render: (inputProps) => <DateTimeInputWithFormValue {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  expect(input.value).toBe('2021-01-15 13:00')
})

test('time is shown in the display time zone if specified (utc+2 summer)', async () => {
  const {result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
      options: {displayTimeZone: 'Europe/Oslo'},
    }),
    props: {documentValue: {test: '2021-06-15T12:00:00.000Z'}},
    render: (inputProps) => <DateTimeInputWithFormValue {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  expect(input.value).toBe('2021-06-15 14:00')
})

test('the time zone can be changed by the user if allowed', async () => {
  const {result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
      options: {displayTimeZone: 'Europe/Oslo', allowTimeZoneSwitch: true},
    }),
    props: {documentValue: {test: '2021-06-15T12:00:00.000Z'}},
    render: (inputProps) => <DateTimeInputWithFormValue {...inputProps} />,
  })

  // click on the TimeZoneButton
  const timeZoneButton = result.getByText('GMT+1')
  userEvent.click(timeZoneButton)
  // ensure the dialog shows
  expect(result.getByText('Select time zone')).toBeInTheDocument()
})

test('the time zone can not be changed by the user if not allowed', async () => {
  const {result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
      options: {displayTimeZone: 'Europe/Oslo', allowTimeZoneSwitch: false},
    }),
    props: {documentValue: {test: '2021-06-15T12:00:00.000Z'}},
    render: (inputProps) => <DateTimeInputWithFormValue {...inputProps} />,
  })

  // click on the TimeZoneButton
  const timeZoneButton = result.getByText('GMT+1')
  userEvent.click(timeZoneButton)
  // ensure the dialog shows
  expect(result.queryByText('Select time zone')).not.toBeInTheDocument()
})
