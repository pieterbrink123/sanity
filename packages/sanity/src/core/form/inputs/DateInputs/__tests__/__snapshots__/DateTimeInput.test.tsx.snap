// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`emits onChange on correct format if a valid value has been typed 1`] = `
[
  [
    undefined,
  ],
]
`;
